diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_m/extmod/extmod.mk bin/m/extmod/extmod.mk
--- bin/ref_m/extmod/extmod.mk	2024-07-21 12:58:41.486695868 -0400
+++ bin/m/extmod/extmod.mk	2024-07-21 12:52:08.990697341 -0400
@@ -174 +174 @@
-OOFATFS_DIR = lib/oofatfs
+OOFATFS_DIR = $(TOP)/lib/oofatfs
@@ -177 +177 @@
-CFLAGS_THIRDPARTY += -DFFCONF_H=\"$(OOFATFS_DIR)/ffconf.h\"
+CFLAGS_THIRDPARTY += -DFFCONF_H=\"$(abspath $(OOFATFS_DIR))/ffconf.h\"
@@ -238,2 +238,2 @@
-MBEDTLS_DIR = lib/mbedtls
-MBEDTLS_CONFIG_FILE ?= \"mbedtls/mbedtls_config_port.h\"
+MBEDTLS_DIR = $(TOP)/lib/mbedtls
+MBEDTLS_CONFIG_FILE ?= \"$(abspath $(PORT_DIR))/mbedtls/mbedtls_config_port.h\"
@@ -242,2 +242,2 @@
-CFLAGS_EXTMOD += -DMICROPY_SSL_MBEDTLS=1 -I$(TOP)/$(MBEDTLS_DIR)/include
-SRC_THIRDPARTY_C += lib/mbedtls_errors/mp_mbedtls_errors.c
+CFLAGS_EXTMOD += -DMICROPY_SSL_MBEDTLS=1 -I$(MBEDTLS_DIR)/include
+SRC_THIRDPARTY_C += $(TOP)/lib/mbedtls_errors/mp_mbedtls_errors.c
@@ -331 +331 @@
-SRC_THIRDPARTY_C += shared/netutils/netutils.c
+SRC_THIRDPARTY_C += $(TOP)/shared/netutils/netutils.c
@@ -384 +384 @@
-BERKELEY_DB_CONFIG_FILE ?= \"extmod/berkeley-db/berkeley_db_config_port.h\"
+BERKELEY_DB_CONFIG_FILE ?= \"$(abspath $(TOP))/extmod/berkeley-db/berkeley_db_config_port.h\"
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_m/mpy-cross/Makefile bin/m/mpy-cross/Makefile
--- bin/ref_m/mpy-cross/Makefile	2024-07-21 12:58:41.754695868 -0400
+++ bin/m/mpy-cross/Makefile	2024-07-21 12:52:08.990697341 -0400
@@ -1 +1 @@
-include ../py/mkenv.mk
+include $(TOP)/py/mkenv.mk
@@ -7 +7 @@
-QSTR_DEFS = qstrdefsport.h
+QSTR_DEFS = $(PORT_DIR)/qstrdefsport.h
@@ -15 +15 @@
-INC += -I.
+INC += -I$(PORT_DIR)
@@ -48 +48 @@
-SRC_C = \
+SRC_CA= \
@@ -51 +51,2 @@
-	shared/runtime/gchelper_generic.c \
+	shared/runtime/gchelper_generic.c 
+
@@ -58,0 +60,2 @@
+SRC_C=$(addprefix $(PORT_DIR)/,$(SRC_CA))
+
@@ -60 +63 @@
-OBJ += $(addprefix $(BUILD)/, $(SRC_C:.c=.o))
+OBJ += $(addprefix $(BUILD)/, $(SRC_CA:.c=.o))
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_m/ports/minimal/Makefile bin/m/ports/minimal/Makefile
--- bin/ref_m/ports/minimal/Makefile	2024-07-21 12:58:41.785695867 -0400
+++ bin/m/ports/minimal/Makefile	2024-07-21 12:52:08.991697341 -0400
@@ -1 +1 @@
-include ../../py/mkenv.mk
+include $(TOP)/py/mkenv.mk
@@ -6 +6 @@
-QSTR_DEFS = qstrdefsport.h
+QSTR_DEFS = $(PORT_DIR)/qstrdefsport.h
@@ -56,2 +56,2 @@
-	main.c \
-	uart_core.c \
+	$(PORT_DIR)/main.c \
+	$(PORT_DIR)/uart_core.c \
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_m/ports/unix/Makefile bin/m/ports/unix/Makefile
--- bin/ref_m/ports/unix/Makefile	2024-07-21 12:58:41.835695867 -0400
+++ bin/m/ports/unix/Makefile	2024-07-21 12:52:08.991697341 -0400
@@ -1,0 +2 @@
+
@@ -9 +10 @@
-VARIANT_DIR ?= variants/$(VARIANT)
+VARIANT_DIR ?= $(PORT_DIR)/variants/$(VARIANT)
@@ -11 +11,0 @@
-
@@ -19,2 +19,2 @@
-include ../../py/mkenv.mk
--include mpconfigport.mk
+include $(TOP)/py/mkenv.mk
+-include $(PORT_DIR)/mpconfigport.mk
@@ -24 +24 @@
-FROZEN_MANIFEST ?= variants/manifest.py
+FROZEN_MANIFEST ?= $(PORT_DIR)/variants/manifest.py
@@ -30 +30 @@
-QSTR_DEFS += qstrdefsport.h
+QSTR_DEFS += $(PORT_DIR)/qstrdefsport.h
@@ -42 +42 @@
-INC +=  -I.
+INC +=  -I$(PORT_DIR)
@@ -197 +197 @@
-SRC_C += \
+SRC_CA +=\
@@ -213,2 +213,3 @@
-	modjni.c \
-	$(wildcard $(VARIANT_DIR)/*.c)
+	modjni.c 
+
+SRC_CA += $(wildcard $(VARIANT_DIR)/*.c)
@@ -216 +217 @@
-SHARED_SRC_C += $(addprefix shared/,\
+SHARED_SRC_CA += $(addprefix shared/,\
@@ -224,2 +225,2 @@
-OBJ = $(PY_O)
-OBJ += $(addprefix $(BUILD)/, $(SRC_C:.c=.o))
+OBJ:= $(PY_O)
+OBJ += $(addprefix $(BUILD)/, $(SRC_CA:.c=.o))
@@ -227 +228,3 @@
-OBJ += $(addprefix $(BUILD)/, $(SHARED_SRC_C:.c=.o))
+OBJ += $(addprefix $(BUILD)/, $(SHARED_SRC_CA:.c=.o))
+SRC_C=$(addprefix $(PORT_DIR)/,$(SRC_CA))
+SHARED_SRC_C:=$(addprefix $(TOP)/,$(SHARED_SRC_CA))
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_m/ports/unix/variants/coverage/mpconfigvariant.mk bin/m/ports/unix/variants/coverage/mpconfigvariant.mk
--- bin/ref_m/ports/unix/variants/coverage/mpconfigvariant.mk	2024-07-21 12:58:41.838695867 -0400
+++ bin/m/ports/unix/variants/coverage/mpconfigvariant.mk	2024-07-21 12:52:08.991697341 -0400
@@ -15,2 +15,2 @@
-SRC_C += coverage.c
-SRC_CXX += coveragecpp.cpp
+SRC_C += $(VARIANT_DIR)/coverage.c
+SRC_CXX += $(PORT_DIR)/coveragecpp.cpp
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_m/py/mkenv.mk bin/m/py/mkenv.mk
--- bin/ref_m/py/mkenv.mk	2024-07-21 12:58:41.847695867 -0400
+++ bin/m/py/mkenv.mk	2024-07-21 12:52:08.991697341 -0400
@@ -13 +13 @@
-TOP := $(patsubst %/py/mkenv.mk,%,$(THIS_MAKEFILE))
+#TOP := $(patsubst %/py/mkenv.mk,%,$(THIS_MAKEFILE))
@@ -35 +35 @@
-BUILD ?= build
+BUILD ?= $(PORT_DIR)/build
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_m/py/mkrules.mk bin/m/py/mkrules.mk
--- bin/ref_m/py/mkrules.mk	2024-07-21 12:58:41.847695867 -0400
+++ bin/m/py/mkrules.mk	2024-07-21 12:58:30.371695891 -0400
@@ -89 +89,4 @@
-$(BUILD)/%.o: %.c
+$(BUILD)/%.o: %.c 
+	$(call compile_c)
+
+$(BUILD)/%.o: $(PORT_DIR)/%.c
@@ -95,0 +99,6 @@
+$(BUILD)/%.o: $(PORT_DIR)/%.cpp
+	$(call compile_cxx)
+
+$(BUILD)/%.o: $(VARIANT_DIR)/%.cpp
+	$(call compile_cxx)
+
@@ -177,0 +187,3 @@
+#$(MAKE) -C "$(abspath $(dir $@)..)"
+EXPORT_VARS:=PORT_DIR TOP MICROPY_MPYCROSS MICROPY_MPYCROSS_DEPENDENCY VARIANT VARIANT_DIR BUILD CROSS_DIR
+EXPORT=$(foreach V, $(EXPORT_VARS),$V=$($V))
@@ -179 +191 @@
-	$(MAKE) -C "$(abspath $(dir $@)..)"
+	$(MAKE) -f $(CROSS_DIR)/Makefile $(EXPORT) PORT_DIR=$(CROSS_DIR) BUILD=$b V=1 2>&1 |tee  $b/mpy_cross_rem.txt
@@ -206 +218 @@
-MICROPY_MANIFEST_PORT_DIR = $(shell pwd)
+MICROPY_MANIFEST_PORT_DIR = $(PORT_DIR)
@@ -214 +226 @@
-$(BUILD)/frozen_content.c: FORCE $(BUILD)/genhdr/qstrdefs.generated.h $(BUILD)/genhdr/root_pointers.h | $(MICROPY_MPYCROSS_DEPENDENCY)
+$(BUILD)/frozen_content.c: $(BUILD)/genhdr/qstrdefs.generated.h $(BUILD)/genhdr/root_pointers.h | $(MICROPY_MPYCROSS_DEPENDENCY)
@@ -216 +228 @@
-	$(Q)$(MAKE_MANIFEST) -o $@ $(MANIFEST_VARIABLES) -b "$(BUILD)" $(if $(MPY_CROSS_FLAGS),-f"$(MPY_CROSS_FLAGS)",) --mpy-tool-flags="$(MPY_TOOL_FLAGS)" $(FROZEN_MANIFEST)
+	$(Q)export MICROPY_MPYCROSS=$(MICROPY_MPYCROSS) && $(MAKE_MANIFEST) -o $@ $(MANIFEST_VARIABLES) -b "$(BUILD)" $(if $(MPY_CROSS_FLAGS),-f"$(MPY_CROSS_FLAGS)",) --mpy-tool-flags="$(MPY_TOOL_FLAGS)" $(FROZEN_MANIFEST)
@@ -254,2 +266,2 @@
-	$(Q)git submodule sync $(addprefix $(TOP)/,$(GIT_SUBMODULES))
-	$(Q)git submodule update --init $(addprefix $(TOP)/,$(GIT_SUBMODULES))
+	$(Q)git -C $(TOP) submodule sync $(GIT_SUBMODULES)
+	$(Q)git -C $(TOP) submodule update --init $(GIT_SUBMODULES)
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_m/py/py.mk bin/m/py/py.mk
--- bin/ref_m/py/py.mk	2024-07-21 12:58:41.853695867 -0400
+++ bin/m/py/py.mk	2024-07-21 12:57:45.289695999 -0400
@@ -18 +18 @@
-QSTR_GLOBAL_DEPENDENCIES += $(PY_SRC)/mpconfig.h mpconfigport.h
+QSTR_GLOBAL_DEPENDENCIES += $(PY_SRC)/mpconfig.h $(PORT_DIR)/mpconfigport.h
@@ -211 +211 @@
-SRC_QSTR += $(filter-out $(SRC_QSTR_IGNORE),$(PY_CORE_O_BASENAME:.o=.c))
+SRC_QSTR += $(addprefix $(TOP)/, $(filter-out $(SRC_QSTR_IGNORE),$(PY_CORE_O_BASENAME:.o=.c)))
@@ -217 +217 @@
-$(HEADER_BUILD)/mpversion.h: FORCE | $(HEADER_BUILD)
+$(HEADER_BUILD)/mpversion.h: | $(HEADER_BUILD)
@@ -230 +230 @@
-$(HEADER_BUILD)/qstrdefs.generated.h: $(PY_QSTR_DEFS) $(QSTR_DEFS) $(QSTR_DEFS_COLLECTED) $(PY_SRC)/makeqstrdata.py mpconfigport.h $(MPCONFIGPORT_MK) $(PY_SRC)/mpconfig.h | $(HEADER_BUILD)
+$(HEADER_BUILD)/qstrdefs.generated.h: $(PY_QSTR_DEFS) $(QSTR_DEFS) $(QSTR_DEFS_COLLECTED) $(PY_SRC)/makeqstrdata.py $(PORT_DIR)/mpconfigport.h $(MPCONFIGPORT_MK) $(PY_SRC)/mpconfig.h | $(HEADER_BUILD)
