diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_micropython_micropython_v1.23.0/extmod/extmod.mk bin/micropython_micropython_v1.23.0/extmod/extmod.mk
--- bin/ref_micropython_micropython_v1.23.0/extmod/extmod.mk	2024-07-20 10:59:26.000000000 -0400
+++ bin/micropython_micropython_v1.23.0/extmod/extmod.mk	2024-07-20 11:46:57.887911268 -0400
@@ -174 +174 @@
-OOFATFS_DIR = lib/oofatfs
+OOFATFS_DIR = $(TOP)/lib/oofatfs
@@ -177 +177 @@
-CFLAGS_THIRDPARTY += -DFFCONF_H=\"$(OOFATFS_DIR)/ffconf.h\"
+CFLAGS_THIRDPARTY += -DFFCONF_H=\"$(abspath $(OOFATFS_DIR))/ffconf.h\"
@@ -238,2 +238,2 @@
-MBEDTLS_DIR = lib/mbedtls
-MBEDTLS_CONFIG_FILE ?= \"mbedtls/mbedtls_config_port.h\"
+MBEDTLS_DIR = $(TOP)/lib/mbedtls
+MBEDTLS_CONFIG_FILE ?= \"$(abspath $(PORT_DIR))/mbedtls/mbedtls_config_port.h\"
@@ -242,2 +242,2 @@
-CFLAGS_EXTMOD += -DMICROPY_SSL_MBEDTLS=1 -I$(TOP)/$(MBEDTLS_DIR)/include
-SRC_THIRDPARTY_C += lib/mbedtls_errors/mp_mbedtls_errors.c
+CFLAGS_EXTMOD += -DMICROPY_SSL_MBEDTLS=1 -I$(MBEDTLS_DIR)/include
+SRC_THIRDPARTY_C += $(TOP)/lib/mbedtls_errors/mp_mbedtls_errors.c
@@ -331 +331 @@
-SRC_THIRDPARTY_C += shared/netutils/netutils.c
+SRC_THIRDPARTY_C += $(TOP)/shared/netutils/netutils.c
@@ -384 +384 @@
-BERKELEY_DB_CONFIG_FILE ?= \"extmod/berkeley-db/berkeley_db_config_port.h\"
+BERKELEY_DB_CONFIG_FILE ?= \"$(abspath $(TOP))/extmod/berkeley-db/berkeley_db_config_port.h\"
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_micropython_micropython_v1.23.0/mpy-cross/Makefile bin/micropython_micropython_v1.23.0/mpy-cross/Makefile
--- bin/ref_micropython_micropython_v1.23.0/mpy-cross/Makefile	2024-07-20 10:59:26.000000000 -0400
+++ bin/micropython_micropython_v1.23.0/mpy-cross/Makefile	2024-07-20 11:46:57.887911268 -0400
@@ -1 +1 @@
-include ../py/mkenv.mk
+include $(TOP)/py/mkenv.mk
@@ -7 +7 @@
-QSTR_DEFS = qstrdefsport.h
+QSTR_DEFS = $(PORT_DIR)/qstrdefsport.h
@@ -15 +15 @@
-INC += -I.
+INC += -I$(PORT_DIR)
@@ -48 +48 @@
-SRC_C = \
+SRC_CA= \
@@ -51 +51,2 @@
-	shared/runtime/gchelper_generic.c \
+	shared/runtime/gchelper_generic.c 
+
@@ -58,0 +60,2 @@
+SRC_C=$(addprefix $(PORT_DIR)/,$(SRC_CA))
+
@@ -60 +63 @@
-OBJ += $(addprefix $(BUILD)/, $(SRC_C:.c=.o))
+OBJ += $(addprefix $(BUILD)/, $(SRC_CA:.c=.o))
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_micropython_micropython_v1.23.0/ports/minimal/Makefile bin/micropython_micropython_v1.23.0/ports/minimal/Makefile
--- bin/ref_micropython_micropython_v1.23.0/ports/minimal/Makefile	2024-07-20 10:59:26.000000000 -0400
+++ bin/micropython_micropython_v1.23.0/ports/minimal/Makefile	2024-07-20 11:46:57.887911268 -0400
@@ -1 +1 @@
-include ../../py/mkenv.mk
+include $(TOP)/py/mkenv.mk
@@ -6 +6 @@
-QSTR_DEFS = qstrdefsport.h
+QSTR_DEFS = $(PORT_DIR)/qstrdefsport.h
@@ -56,2 +56,2 @@
-	main.c \
-	uart_core.c \
+	$(PORT_DIR)/main.c \
+	$(PORT_DIR)/uart_core.c \
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_micropython_micropython_v1.23.0/ports/unix/Makefile bin/micropython_micropython_v1.23.0/ports/unix/Makefile
--- bin/ref_micropython_micropython_v1.23.0/ports/unix/Makefile	2024-07-20 10:59:26.000000000 -0400
+++ bin/micropython_micropython_v1.23.0/ports/unix/Makefile	2024-07-20 11:46:57.887911268 -0400
@@ -1,0 +2 @@
+
@@ -9 +10 @@
-VARIANT_DIR ?= variants/$(VARIANT)
+VARIANT_DIR ?= $(PORT_DIR)/variants/$(VARIANT)
@@ -11 +11,0 @@
-
@@ -19,2 +19,2 @@
-include ../../py/mkenv.mk
--include mpconfigport.mk
+include $(TOP)/py/mkenv.mk
+-include $(PORT_DIR)/mpconfigport.mk
@@ -24 +24 @@
-FROZEN_MANIFEST ?= variants/manifest.py
+FROZEN_MANIFEST ?= $(PORT_DIR)/variants/manifest.py
@@ -30 +30 @@
-QSTR_DEFS += qstrdefsport.h
+QSTR_DEFS += $(PORT_DIR)/qstrdefsport.h
@@ -42 +42 @@
-INC +=  -I.
+INC +=  -I$(PORT_DIR)
@@ -197 +197 @@
-SRC_C += \
+SRC_CA +=\
@@ -213,2 +213,3 @@
-	modjni.c \
-	$(wildcard $(VARIANT_DIR)/*.c)
+	modjni.c 
+
+SRC_CA += $(wildcard $(VARIANT_DIR)/*.c)
@@ -216 +217 @@
-SHARED_SRC_C += $(addprefix shared/,\
+SHARED_SRC_CA += $(addprefix shared/,\
@@ -224,2 +225,2 @@
-OBJ = $(PY_O)
-OBJ += $(addprefix $(BUILD)/, $(SRC_C:.c=.o))
+OBJ:= $(PY_O)
+OBJ += $(addprefix $(BUILD)/, $(SRC_CA:.c=.o))
@@ -227 +228,3 @@
-OBJ += $(addprefix $(BUILD)/, $(SHARED_SRC_C:.c=.o))
+OBJ += $(addprefix $(BUILD)/, $(SHARED_SRC_CA:.c=.o))
+SRC_C=$(addprefix $(PORT_DIR)/,$(SRC_CA))
+SHARED_SRC_C:=$(addprefix $(TOP)/,$(SHARED_SRC_CA))
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_micropython_micropython_v1.23.0/ports/unix/variants/coverage/mpconfigvariant.mk bin/micropython_micropython_v1.23.0/ports/unix/variants/coverage/mpconfigvariant.mk
--- bin/ref_micropython_micropython_v1.23.0/ports/unix/variants/coverage/mpconfigvariant.mk	2024-07-20 10:59:26.000000000 -0400
+++ bin/micropython_micropython_v1.23.0/ports/unix/variants/coverage/mpconfigvariant.mk	2024-07-20 11:51:49.452910003 -0400
@@ -15,2 +15,2 @@
-SRC_C += coverage.c
-SRC_CXX += coveragecpp.cpp
+SRC_C += $(VARIANT_DIR)/coverage.c
+SRC_CXX += $(PORT_DIR)/coveragecpp.cpp
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_micropython_micropython_v1.23.0/py/mkenv.mk bin/micropython_micropython_v1.23.0/py/mkenv.mk
--- bin/ref_micropython_micropython_v1.23.0/py/mkenv.mk	2024-07-20 10:59:26.000000000 -0400
+++ bin/micropython_micropython_v1.23.0/py/mkenv.mk	2024-07-20 11:46:57.887911268 -0400
@@ -13 +13 @@
-TOP := $(patsubst %/py/mkenv.mk,%,$(THIS_MAKEFILE))
+#TOP := $(patsubst %/py/mkenv.mk,%,$(THIS_MAKEFILE))
@@ -35 +35 @@
-BUILD ?= build
+BUILD ?= $(PORT_DIR)/build
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_micropython_micropython_v1.23.0/py/mkrules.mk bin/micropython_micropython_v1.23.0/py/mkrules.mk
--- bin/ref_micropython_micropython_v1.23.0/py/mkrules.mk	2024-07-20 10:59:26.000000000 -0400
+++ bin/micropython_micropython_v1.23.0/py/mkrules.mk	2024-07-20 11:51:21.454910134 -0400
@@ -89 +89,4 @@
-$(BUILD)/%.o: %.c
+$(BUILD)/%.o: %.c 
+	$(call compile_c)
+
+$(BUILD)/%.o: $(PORT_DIR)/%.c
@@ -95,0 +99,6 @@
+$(BUILD)/%.o: $(PORT_DIR)/%.cpp
+	$(call compile_cxx)
+
+$(BUILD)/%.o: $(VARIANT_DIR)/%.cpp
+	$(call compile_cxx)
+
@@ -206 +215 @@
-MICROPY_MANIFEST_PORT_DIR = $(shell pwd)
+MICROPY_MANIFEST_PORT_DIR = $(PORT_DIR)
@@ -254,2 +263,2 @@
-	$(Q)git submodule sync $(addprefix $(TOP)/,$(GIT_SUBMODULES))
-	$(Q)git submodule update --init $(addprefix $(TOP)/,$(GIT_SUBMODULES))
+	$(Q)git -C $(TOP) submodule sync $(GIT_SUBMODULES)
+	$(Q)git -C $(TOP) submodule update --init $(GIT_SUBMODULES)
diff -U0 -rN '--exclude=*.git*' '--exclude=__pycache__' bin/ref_micropython_micropython_v1.23.0/py/py.mk bin/micropython_micropython_v1.23.0/py/py.mk
--- bin/ref_micropython_micropython_v1.23.0/py/py.mk	2024-07-20 10:59:26.000000000 -0400
+++ bin/micropython_micropython_v1.23.0/py/py.mk	2024-07-20 11:46:57.887911268 -0400
@@ -18 +18 @@
-QSTR_GLOBAL_DEPENDENCIES += $(PY_SRC)/mpconfig.h mpconfigport.h
+QSTR_GLOBAL_DEPENDENCIES += $(PY_SRC)/mpconfig.h $(PORT_DIR)/mpconfigport.h
@@ -211 +211 @@
-SRC_QSTR += $(filter-out $(SRC_QSTR_IGNORE),$(PY_CORE_O_BASENAME:.o=.c))
+SRC_QSTR += $(addprefix $(TOP)/, $(filter-out $(SRC_QSTR_IGNORE),$(PY_CORE_O_BASENAME:.o=.c)))
@@ -230 +230 @@
-$(HEADER_BUILD)/qstrdefs.generated.h: $(PY_QSTR_DEFS) $(QSTR_DEFS) $(QSTR_DEFS_COLLECTED) $(PY_SRC)/makeqstrdata.py mpconfigport.h $(MPCONFIGPORT_MK) $(PY_SRC)/mpconfig.h | $(HEADER_BUILD)
+$(HEADER_BUILD)/qstrdefs.generated.h: $(PY_QSTR_DEFS) $(QSTR_DEFS) $(QSTR_DEFS_COLLECTED) $(PY_SRC)/makeqstrdata.py $(PORT_DIR)/mpconfigport.h $(MPCONFIGPORT_MK) $(PY_SRC)/mpconfig.h | $(HEADER_BUILD)
